### PR TITLE
Training function that is called fine_tune

### PR DESCRIPTION
**Pull request by lorenzoCapelli**
I modified the function fine_tune adding some stuff

### CODE REVIEW

File: peepholelib/utils/fine_tune.py [hunk 1]
```diff
@@ -4,30 +4,52 @@
...
+        requires_grad: Whether to enable gradients
+        layer_names: Optional list of layer names to target. If None, affects all parameters.
+    """
+    if layer_names is None:
+        # Affect all parameters
+        for param in model.parameters():
+            param.requires_grad = requires_grad
+    else:
+        # Affect only specified layers
+        for name, param in model.named_parameters():
```
Comment (lsrosa) [inline peepholelib/utils/fine_tune.py]: This double loop seems unecessary.

File: peepholelib/utils/fine_tune.py [hunk 2]
```diff
@@ -4,30 +4,52 @@
...
 from torch.utils.data import DataLoader
-from torch.nn import DataParallel
 
 def img_classification_acc(pred, target):
-    return torch.count_nonzero(torch.argmax(pred, axis=1)==labels)
+    
+    pred_idx = torch.argmax(pred, dim=1)
+    return (pred_idx == target).sum()
+
+def set_requires_grad(model, requires_grad, layer_names):
```
Comment (lsrosa) [inline peepholelib/utils/fine_tune.py]: Use kwargs and default values

File: peepholelib/utils/fine_tune.py [hunk 3]
```diff
@@ -39,20 +61,47 @@ def fine_tune(**kwargs):
...
     # training progress
-    lr = kwargs['lr']
-    patience = kwargs.get('patience', 10)
-    iterations = kwargs.get('iterations', 'full')
+    lr = kwargs.get('lr', 5e-5)
     bs = kwargs.get('batch_size', 256)
-    max_epochs = kwargs.get('max_epochs', 1e3)
+    max_epochs = kwargs.get('max_epochs', 1000)
+    iterations = kwargs.get('iterations', 'full')
+    n_threads = kwargs.get('n_threads', 0)
```
Comment (lsrosa) [inline peepholelib/utils/fine_tune.py:49]: n_threads defaults to 0. Does this work?

File: peepholelib/utils/fine_tune.py [hunk 4]
```diff
@@ -39,20 +61,47 @@ def fine_tune(**kwargs):
...
     # create training artifacts
     loss_fn = _l(**loss_kwargs)
-    optim = _opt(model._model.parameters(), lr=lr, **optim_kwargs)
-    scheduler = _sched(optimizer=optim, **scheduler_kwargs) if not _sched == None else None
 
-    # saving
-    save_every = kwargs['save_every'] if 'save_every' in kwargs else 100
-    verbose = kwargs['verbose'] if 'verbose' in kwargs else False
+    # Apply initial layer freezing
+    if freeze_all_but is not None:
```
Comment (lsrosa) [inline peepholelib/utils/fine_tune.py:67]: `freeze_all_but` interface could be the same as `set_requires_grad` for uniformity.

File: peepholelib/utils/fine_tune.py [hunk 5]
```diff
@@ -95,43 +156,61 @@ def fine_tune(**kwargs):
...
 
-        initial_epoch = trained_for 
+                same_state = current_state == saved_state
+                if same_state: scheduler.load_state_dict(data['scheduler'])
+          
     else:
         if verbose: print('No training ongoing, starting anew.')
         initial_epoch = 0
+        best_val_loss = float('inf')
+        best_epoch = None
```
Comment (lsrosa) [inline peepholelib/utils/fine_tune.py]: should  `best_epoch` start at 0?

File: peepholelib/utils/fine_tune.py [hunk 6]
```diff
@@ -95,43 +156,61 @@ def fine_tune(**kwargs):
...
                 name = _f,
                 vebose = verbose
                 )
-        optim.load_state_dict(data['optimizer']) 
-        if not scheduler == None:
-            scheduler.load_state_dict(data['scheduler'])
+        
+        initial_epoch = trained_for
+        
+        if data['freeze_all_but'] == freeze_all_but:
```
Comment (lsrosa) [inline peepholelib/utils/fine_tune.py:151]: Seems like if `freeze_all_but` is not passed, the optimizer's and schedule's state will not be loaded. I might have missed something.

File: peepholelib/utils/fine_tune.py [hunk 7]
```diff
@@ -95,43 +156,61 @@ def fine_tune(**kwargs):
...
+        
+            optim.load_state_dict(data['optimizer']) 
+
+            if scheduler is not None and 'scheduler' in data and data['scheduler'] is not None:
```
Comment (lsrosa) [inline peepholelib/utils/fine_tune.py:155]: I do not get why all this data check, feels like it is solving a bug I was unaware of.

File: peepholelib/utils/fine_tune.py (hunk 2 above)
Comment (lsrosa) [inline peepholelib/utils/fine_tune.py]: Should this `set_requires_grad` be moved inside `ModelWrap`?

File: peepholelib/peepholes/classifiers/tkmeans.py [hunk 8]
```diff
@@ -104,7 +104,7 @@ def save(self, **kwargs):
 
     def load(self, **kwargs):
         if self._clas_path.exists(): 
-            self._classifier = tGMM.load(self._clas_path)
+            self._classifier = tKMeans.load(self._clas_path)
```
Comment (lsrosa) [inline peepholelib/peepholes/classifiers/tkmeans.py:107]: nice catch

**Review by lsrosa — CHANGES_REQUESTED** [review]

File: peepholelib/utils/fine_tune.py (hunk 1 above)
Comment (lorenzoCapelli) [reply peepholelib/utils/fine_tune.py]: i didn't get what you mean here

File: peepholelib/utils/fine_tune.py (hunk 3 above)
Comment (lorenzoCapelli) [reply peepholelib/utils/fine_tune.py:49]: yes it does idk why

File: peepholelib/utils/fine_tune.py (hunk 4 above)
Comment (lorenzoCapelli) [reply peepholelib/utils/fine_tune.py:67]: Also here i am not getting what you mean

File: peepholelib/utils/fine_tune.py (hunk 7 above)
Comment (lorenzoCapelli) [reply peepholelib/utils/fine_tune.py:155]: The problem is that if the configuration is the same a the previous call of the training we load what we had before otherwise we have to change stuff

File: peepholelib/utils/fine_tune.py (hunk 1 above)
Comment (lsrosa) [reply peepholelib/utils/fine_tune.py]: Lines 33 and 34 are iterating over all `layer_names` and all `namede_params`, it should be possible to access the module (layers) directly from its names. If I am not mistaken using `modelWrap._model._modules`, so you do not need to iterate over all model's parameters.

File: peepholelib/utils/fine_tune.py (hunk 4 above)
Comment (lsrosa) [reply peepholelib/utils/fine_tune.py:67]: `freeze_all_but` only takes the name of the layer to train, while `set_requires_grad` takes many layers to train. It is better to keep the interface of the latter since it is more generic and implements the first. So `fine_tune()` should receive an argument like `layers_to_train` instead of `freeze_all_but`, and pass that to  `set_requires_grad`.

File: peepholelib/utils/fine_tune.py [hunk 9]
```diff
@@ -95,43 +156,61 @@ def fine_tune(**kwargs):
...
 
     for epoch in range(initial_epoch, max_epochs):
+        
         t0 = time()
+
         # peform train iterations
         loss_acc = 0.0
         acc_acc = 0.0
         samples_acc = 0
+        epochs_without_improvement = 0
```
Comment (lsrosa) [inline peepholelib/utils/fine_tune.py:185]: Just a friendly discussion:

This thing about manually counting the epochs for early stopping sort of bothers me. This number is used by the scheduler to reduce the learning rate, and can create a conflict with early stopping.

If I understood correctly, the manual early stopping works even without the scheduler, which is nice though. 
My initial idea was that early stopping was not necessary since we are saving checkpoints, so we could leave the scheduler do its thing, which is the common approach, but indeed it is nice to have some criterion for that.

One suggestion is to remove this `epochs_without_improvement` and use the scheduler's ` self.num_bad_epochs` (see the [schedule's code](https://github.com/pytorch/pytorch/blob/v2.10.0/torch/optim/lr_scheduler.py#L1691)) which implements the very same checks, and add a check in the beginning of `fine_tune` to assert that `early_stopping_patiance > scheduler.patience` so both will not conflict. This would make the scheduler required for the early patience, which is less flexible than having both independent from each other, but that should be the scheduler's job, so I do not see many problems in forcing them together.

File: peepholelib/utils/fine_tune.py (hunk 2 above)
Comment (lsrosa) [reply peepholelib/utils/fine_tune.py]: Now that I double think, this should be in the `ModelWrap` and a we should make a first call setting all modules to `set_requires_grad = False` at  `ModelWrap`'s initialization.

File: peepholelib/models/model_wrap.py [hunk 10]
```diff
@@ -164,6 +164,28 @@ def __call__(self, x):
 
         return res 
     
+    def set_requires_grad(self, **kwargs):
+        """
+        Set requires_grad for model parameters.
+        
+        Args:
+            model: PyTorch model
```
Comment (lsrosa) [inline peepholelib/models/model_wrap.py:172]: clean model from docs

File: peepholelib/utils/fine_tune.py (hunk 1 above)
Comment (lorenzoCapelli) [reply peepholelib/utils/fine_tune.py]: I don't think it is the same, i tested on ViT and `modelWrap._model._modules` returns me just the keys of the macro modules and not the single layers while  `model.named_parameters()` provides me the full list of layers and their parameters

File: peepholelib/models/model_wrap.py [hunk 11]
```diff
@@ -178,39 +200,63 @@ def update_output(self, **kwargs):
...
 
-        else:
-            out_size = temp[-1].out_features 
-            temp.append(torch.nn.Linear(out_size, n_classes, device=self.device))
+        n_classes = kwargs['to_n_classes']
 
-        return
+        model = self._model
+
+        for attr in ["fc", "classifier", "heads"]:
```
Comment (lsrosa) [inline peepholelib/models/model_wrap.py:232]: this is very specific for the models we are testing. It is not a good practice to use `attr` since they are arbitrary strings (each model will have its own), this will also not work if the model has a submodule (layer) with the names in the list. We should get a direct reference for the model's output.

**Review by lsrosa — CHANGES_REQUESTED** [review]

File: peepholelib/utils/fine_tune.py (hunk 4 above)
Comment (lorenzoCapelli) [reply peepholelib/utils/fine_tune.py:67]: ok got it

File: peepholelib/utils/fine_tune.py (hunk 2 above)
Comment (lorenzoCapelli) [reply peepholelib/utils/fine_tune.py]: ok

File: peepholelib/utils/fine_tune.py (hunk 9 above)
Comment (lorenzoCapelli) [reply peepholelib/utils/fine_tune.py:185]: ok i chack that
